### PR TITLE
Add SAM3 video model to the zoo

### PR DESCRIPTION
**Pull request by manushreegangwar**
## 🔗 Related Issues

No related issues to link

## 📋 What changes are proposed in this pull request?

This PR adds SAM3 video model the model zoo. The integration supports concept prompting as well as visual prompting.

## 🧪 How is this patch tested? If it is not, please explain why.

``` 
$ pytest tests/intensive/sam_image_model_tests.py --timeout=500 -v
$ pytest tests/intensive/sam_parity_tests.py --timeout=500 -v
$ pytest tests/intensive/sam_video_model_tests.py --timeout=500 -v
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

SAM3 video model with concept and visual prompting added to the zoo.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other

## Summary by CodeRabbit

* **New Features**
  * Added SAM3 video support with concept (text) and visual (box/keypoint) prompting, configurable propagation settings, prompt-frame selection, per-frame mask+box outputs, and prompt→label mapping.
  * Enhanced SAM3 video processing with FiftyOne-based frame loading and normalization.

* **Tests**
  * Added intensive SAM3 video concept/visual integration tests ensuring detections are populated across frames (including mask presence in concept mode).
  * Expanded SAM parity coverage for SAM3 video (forward/backward/both propagation and keypoint prompting) and strengthened GPU/resource cleanup for deterministic runs.

---
**Enterprise sync PR**: https://github.com/voxel51/fiftyone-teams/pull/2863

### CODE REVIEW

File: fiftyone/utils/sam3.py [hunk 1]
```diff
@@ -545,3 +551,742 @@ def _forward_pass_auto(self, imgs):
...
+        frame_width: original frame width
+        frame_height: original frame height
+        label_map: optional dict mapping obj_id -> label string
+        default_label: fallback label when ``label_map`` is ``None``
+
+    Returns:
+        list of :class:`fiftyone.core.labels.Detection`
+    """
+    detections = []
+    for i, obj_id in enumerate(out_obj_ids):
```
Comment (kaixi-wang) [inline fiftyone/utils/sam3.py:645]: nit: instead of looping, couldn't you vectorize this?

Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:645]: vectorized the bounding box operation

File: fiftyone/utils/sam3.py [hunk 2]
```diff
@@ -545,3 +551,740 @@ def _forward_pass_auto(self, imgs):
...
+            _low_res_masks,
+            out_mask_logits,
+            _obj_scores,
+        ) in self.visual_predictor.propagate_in_video(
+            inference_state,
+            start_frame_idx=None,
+            max_frame_num_to_track=None,
+            reverse=False,
+            propagate_preflight=True,
+        ):
```
[the remark this answers is not in the document]
Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:1361]: fixed

File: tests/intensive/sam_video_model_tests.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        """All concept-mode detections should have instance masks."""
+        field = "sam3v_concept_masks"
+        self.dataset.apply_model(self.model, label_field=field)
+
+        for sample in self.dataset.iter_samples(progress=False):
+            for frame_number, frame in sample.frames.items():
+                dets = frame.get_field(field)
+                if dets is None:
+                    continue
+                for det in dets.detections:
```
[the remark this answers is not in the document]
Comment (manushreegangwar) [reply tests/intensive/sam_video_model_tests.py:106]: added a not none check for detections on frame

File: fiftyone/utils/sam3.py [hunk 4]
```diff
@@ -545,3 +551,820 @@ def _forward_pass_auto(self, imgs):
...
+            seen_indices = set()
+            combined = []
+            for det in fwd_dets.detections:
+                combined.append(det)
+                if det.index is not None:
+                    seen_indices.add(det.index)
+
+            for det in bwd_dets.detections:
+                if det.index is None or det.index not in seen_indices:
+                    combined.append(det)
```
Comment (neerajaabhyankar) [inline fiftyone/utils/sam3.py:1431]: Looks like we're getting rid of all backward detections if a forward one exists of the same index.
Is this what users may expect by default? Maybe we can have a `merge_method` flag that can switch between forward-wins or [max-prediction like here](https://github.com/voxel51/fiftyone-labs/blob/main/plugins/label_propagation/propagation.py#L162) or something confidence-based (at the risk of overcomplicating it).

File: fiftyone/utils/sam3.py
Comment (neerajaabhyankar) [inline fiftyone/utils/sam3.py:336]: [question, unrelated to this PR] just like in SAM2, why don't we need to specify a checkpoint path etc. over here?

I had to add something like this to make it look for the model in the right place (the fo config model_zoo_dir) as opposed to python site-packages.
```
        if "checkpoint_path" not in config.entrypoint_args:
            config.entrypoint_args["checkpoint_path"] = config.model_path
        if "bpe_path" not in config.entrypoint_args:
            config.entrypoint_args["bpe_path"] = config.model_path.replace("sam3.pt", "bpe_simple_vocab_16e6.txt.gz")
```

**Review by neerajaabhyankar — APPROVED** [review]
Looked through the code and was able to run with this -- works great 🙌 no other issues.

File: fiftyone/utils/sam3.py
Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:336]: https://github.com/voxel51/fiftyone/pull/7515 makes the `bpe_path` configurable.

File: fiftyone/utils/sam3.py (hunk 4 above)
Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:1431]: Just keeping it simple -- inject backward detections when forward are missing. Don't have a signal for what users expect. Did you notice any instances where this was inferior to the fusion method? I will add an option to configure backward or forward as the primary detection.

File: fiftyone/utils/sam3.py
Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:336]: Merged the PR and added the bpe related changes to video model as well. Fixed model loading as well.

File: fiftyone/utils/sam3.py (hunk 4 above)
Comment (manushreegangwar) [reply fiftyone/utils/sam3.py:1431]: Added this configuration to model config

**Comment by manushreegangwar** [comment]
> Looked through the code and was able to run with this -- works great 🙌 no other issues.

Thanks for the review, @neerajaabhyankar! Could you look at the updates before I merge this?

File: fiftyone/utils/sam3.py (hunk 4 above)
Comment (neerajaabhyankar) [reply fiftyone/utils/sam3.py:1431]: > Did you notice any instances where this was inferior to the fusion method

Rarely, for very occlusion-heavy datasets, although this should work for now since it's interpretable. Always-fusing is also not guaranteed to be foolproof.

**Review by neerajaabhyankar — APPROVED** [review]
tested ✅ thanks for the updates!
